### PR TITLE
relax strict dtype checking in ecsv to a warning

### LINKED ISSUE #12840: No longer able to read BinnedTimeSeries with datetime column saved as ECSV after upgrading from 4.2.1 -> 5.0+
### Description

Hi, [This commit](https://github.com/astropy/astropy/commit/e807dbff9a5c72bdc42d18c7d6712aae69a0bddc) merged in PR #11569 breaks my ability to read an ECSV file created using Astropy v 4.2.1, BinnedTimeSeries class's write method, which has a datetime64 column. Downgrading astropy back to 4.2.1 fixes the issue because the strict type checking in line 177 of ecsv.py is not there.

Is there a reason why this strict type checking was added to ECSV? Is there a way to preserve reading and writing of ECSV files created with BinnedTimeSeries across versions? I am happy to make a PR on this if the strict type checking is allowed to be scaled back or we can add datetime64 as an allowed type. 

### Expected behavior

The file is read into a `BinnedTimeSeries` object from ecsv file without error.

### Actual behavior

ValueError is produced and the file is not read because ECSV.py does not accept the datetime64 column.
`ValueError: datatype 'datetime64' of column 'time_bin_start' is not in allowed values ('bool', 'int8', 'int16', 'int32', 'int64', 'uint8', 'uint16', 'uint32', 'uint64', 'float16', 'float32', 'float64', 'float128', 'string')`

### Steps to Reproduce

The file is read using:    
`BinnedTimeSeries.read('<file_path>', format='ascii.ecsv')`
which gives a long error. 

The file in question is a binned time series created by  `astropy.timeseries.aggregate_downsample`. which itself is a binned version of an `astropy.timeseries.TimeSeries` instance with some TESS data. (loaded via TimeSeries.from_pandas(Tess.set_index('datetime')). I.e., it has a datetime64 index.  The file was written using the classes own .write method in Astropy V4.2.1 from an instance of said class:   
`myBinnedTimeSeries.write('<file_path>',format='ascii.ecsv',overwrite=True)`

I'll attach a concatenated version of the file (as it contains private data). However, the relevant part from the header is on line 4:

```
# %ECSV 0.9
# ---
# datatype:
# - {name: time_bin_start, datatype: datetime64}
```

as you can see, the datatype is datetime64. This works fine with ECSV V0.9 but not V1.0 as some sort of strict type checking was added. 

### 
Full error log:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [3], in <module>
---> 49 tsrbin = BinnedTimeSeries.read('../Photometry/tsr_bin.dat', format='ascii.ecsv')

File ~/Apps/miniconda3/envs/py310_latest/lib/python3.10/site-packages/astropy/timeseries/binned.py:285, in BinnedTimeSeries.read(self, filename, time_bin_start_column, time_bin_end_column, time_bin_size_column, time_bin_size_unit, time_format, time_scale, format, *args, **kwargs)
    230 """
    231 Read and parse a file and returns a `astropy.timeseries.BinnedTimeSeries`.
    232 
   (...)
    279 
    280 """
    282 try:
    283 
    284     # First we try the readers defined for the BinnedTimeSeries class
--> 285     return super().read(filename, format=format, *args, **kwargs)
    287 except TypeError:
    288 
    289     # Otherwise we fall back to the default Table readers
    291     if time_bin_start_column is None:

File ~/Apps/miniconda3/envs/py310_latest/lib/python3.10/site-packages/astropy/table/connect.py:62, in TableRead.__call__(self, *args, **kwargs)
     59 units = kwargs.pop('units', None)
     60 descriptions = kwargs.pop('descriptions', None)
---> 62 out = self.registry.read(cls, *args, **kwargs)
     64 # For some readers (e.g., ascii.ecsv), the returned `out` class is not
     65 # guaranteed to be the same as the desired output `cls`.  If so,
     66 # try coercing to desired class without copying (io.registry.read
     67 # would normally do a copy).  The normal case here is swapping
     68 # Table <=> QTable.
     69 if cls is not out.__class__:

File ~/Apps/miniconda3/envs/py310_latest/lib/python3.10/site-packages/astropy/io/registry/core.py:199, in UnifiedInputRegistry.read(self, cls, format, cache, *args, **kwargs)
    195     format = self._get_valid_format(
    196         'read', cls, path, fileobj, args, kwargs)
    198 reader = self.get_reader(format, cls)
--> 199 data = reader(*args, **kwargs)
    201 if not isinstance(data, cls):
    202     # User has read with a subclass where only the parent class is
    203     # registered.  This returns the parent class, so try coercing
    204     # to desired subclass.
    205     try:

File ~/Apps/miniconda3/envs/py310_latest/lib/python3.10/site-packages/astropy/io/ascii/connect.py:18, in io_read(format, filename, **kwargs)
     16     format = re.sub(r'^ascii\.', '', format)
     17     kwargs['format'] = format
---> 18 return read(filename, **kwargs)

File ~/Apps/miniconda3/envs/py310_latest/lib/python3.10/site-packages/astropy/io/ascii/ui.py:376, in read(table, guess, **kwargs)
    374     else:
    375         reader = get_reader(**new_kwargs)
--> 376         dat = reader.read(table)
    377         _read_trace.append({'kwargs': copy.deepcopy(new_kwargs),
    378                             'Reader': reader.__class__,
    379                             'status': 'Success with specified Reader class '
    380                                       '(no guessing)'})
    382 # Static analysis (pyright) indicates `dat` might be left undefined, so just
    383 # to be sure define it at the beginning and check here.

File ~/Apps/miniconda3/envs/py310_latest/lib/python3.10/site-packages/astropy/io/ascii/core.py:1343, in BaseReader.read(self, table)
   1340 self.header.update_meta(self.lines, self.meta)
   1342 # Get the table column definitions
-> 1343 self.header.get_cols(self.lines)
   1345 # Make sure columns are valid
   1346 self.header.check_column_names(self.names, self.strict_names, self.guessing)

File ~/Apps/miniconda3/envs/py310_latest/lib/python3.10/site-packages/astropy/io/ascii/ecsv.py:177, in EcsvHeader.get_cols(self, lines)
    175 col.dtype = header_cols[col.name]['datatype']
    176 if col.dtype not in ECSV_DATATYPES:
--> 177     raise ValueError(f'datatype {col.dtype!r} of column {col.name!r} '
    178                      f'is not in allowed values {ECSV_DATATYPES}')
    180 # Subtype is written like "int64[2,null]" and we want to split this
    181 # out to "int64" and [2, None].
    182 subtype = col.subtype

ValueError: datatype 'datetime64' of column 'time_bin_start' is not in allowed values ('bool', 'int8', 'int16', 'int32', 'int64', 'uint8', 'uint16', 'uint32', 'uint64', 'float16', 'float32', 'float64', 'float128', 'string')
```
### System Details

(For the version that does not work)
Python 3.10.2 | packaged by conda-forge | (main, Feb  1 2022, 19:28:35) [GCC 9.4.0]
Numpy 1.22.2
pyerfa 2.0.0.1
astropy 5.0.1
Scipy 1.8.0
Matplotlib 3.5.1

(For the version that does work)
Python 3.7.11 (default, Jul 27 2021, 14:32:16) [GCC 7.5.0]
Numpy 1.20.3
pyerfa 2.0.0.1
astropy 4.2.1
Scipy 1.7.0
Matplotlib 3.4.2

### LINKED ISSUE #12885: Change ECSV version 1.0 datatype checking to a warning.
### Description

This issue is a follow-up to #12840 and #12880. As proposed in #12841, ECSV should issue a `warning` instead of an error for non-standard data types. As this is an API change, it would wait until `Astropy` 5.1. 

Currently, non-standard datatypes can only be read if the header is the old `ECSV V0.9` format. Although non-standard datatypes are not supported in this manner by the spec, there is no reason to limit their use even with ECSV v1.0 files, as long as an appropriate warning is provided. As ECSV is not just an astropy spec, and there seems to be non-compliant files out there in use, it cannot hurt being relatively flexible. 

- [x] Modify 12841 to not have any conflicts with #12880.
- [x] Ensure we test non-compliant edge cases and get a predictable result. (E.g., One that fails, and one that succeeds).
- [ ] Document new behavior.

### Additional context

The proper way of serializing complex objects is written in the spec, documentation, and already exists in the code, so this is independent of that (which uses meta information in the header). This is specifically for datatypes like `str` or `object` or various `numpy dtypes`. Additionally, the specific errors in #12840 are being worked on separately to be compliant with ECSV spec (#12842).

### PR DESCRIPTION
**Pull request by emirkmo**
### Description

**Edit: Needs to wait for #12880 due to likely Merge conflicts.**

This pull request is to address strict type checking in ecsv.py which prevents `BinnedTimeSeries` (and likely other) classes from reading and writing non-standard dtypes to and from an ecsv file. Instead of raising an exception, we issue a `UserWarning` instead, and attempt to go through the file reading. If this fails, the supported dtypes are reported in the warning so the user may switch to them instead.

Also updated tests to expect a warning but I am not sure if this was the right way to approach it. (`test_read_bad_datatype`, `test_read_complex`). 

`UserWarning` text is now:   
`(f'unexpected datatype {col.dtype!r} of column {col.name!r} '
                       f'is not in default datatypes {ECSV_DATATYPES}, using anyway.')`

Edit: Fixes #12885 
~~Fixes #12840~~

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by emirkmo** [comment]
@taldcroft Tagging you here to show my proposal. Let me know if this is not what you had in mind or I am missing something. Don't want to break the much more widely used Table.read functionality on accident.

**Comment by taldcroft** [comment]
Excellent this is on the right track. I'll try to get back later for a review.

**Comment by emirkmo** [comment]
Testing failures seem unrelated, might be due to #12823 . At least I hope.

**Comment by taldcroft** [comment]
The only quick thought I had was whether it would be worth making a subclass like `InvalidEcsvDatatype(AstropyUserWarning)` that would make it easier to squelch this warning globally.

**Comment by emirkmo** [comment]
I thought to make a custom warning but didn’t see many of them in the warnings module, so figured it wasn’t widely used in astropy. Can we make the warning inside ecsv.py by inheriting from UserWarning or should it go into to warnings module?

**Comment by emirkmo** [comment]
@taldcroft Okay, I added InvalidEcsvDatatypeWarning as and changed the tests. Let me know if this is looking alright to you.

**Comment by taldcroft** [comment]
Also I'd request squashing down to one commit since this is a pretty small and contained change.

**Comment by emirkmo** [comment]
Ok, I think I squashed but I wasn't really sure how to do it with your commit having the credentials. please let me know if it looks ok. There are testing failures in py3.9 related to warnings and exceptions having None I think. Thanks for the assistance!

**Comment by taldcroft** [comment]
@emirkmo - something we wrong with CI. This may be fixed by rebasing on main and force pushing. Thanks for your patience...

**Comment by emirkmo** [comment]
@taldcroft Ok, I rebased. I also added a test for str instead of str raising a warning, but feel free to reject that commit or let me know if I should drop it (or rebase it). Like I said, tests are passing locally for me, but I only re-tested the io.ascii package

**Comment by emirkmo** [comment]
Alright, I found the error. Older versions of numpy (`1.18.*`) do not support giving `'complex'` as a datatype string. Needs to be `1.19.*`

```
vals = ['1+1j', '2+2j']
numpy.array(vals, 'complex')
```
fails on numpy=1.18.1 and 1.18.5, but passes on >=1.19.* which is in the tests. And this comes from `test_read_complex()` because we now give a warning and proceed instead of raising an error. Should the tests be modified to expect a warning AND a (possible) error?

**Comment by emirkmo** [comment]
@taldcroft All tests passing now, so we can merge, I hope! 

See my comment above re: tests. I'm also squashing to 1 commit as you asked and changing changelog from bugfix to api change. (will re-run CI, but it already passed before the squash).

While I wish this would go in as a bugfix I see why it should wait for next minor version. Thanks for the help.

Finally, possibly, this behavior should be documented in the [docs](https://docs.astropy.org/en/stable/io/ascii/ecsv.html) somewhere.

**Comment by pllim** [comment]
#12880 is merged. Please rebase. Thanks!

**Comment by emirkmo** [comment]
I put my proposal under a new commit because I think the warning should be given even for V0.9 files.

I also simplified the tests as a result, removing V0.9 specific versions, and adding some asserts to check that the output after proceeding with warning is consistent with what happened with V0.9 (copying from the V0.9 tests)

File: astropy/io/ascii/tests/test_ecsv.py [hunk 1]
```diff
@@ -631,94 +629,80 @@ def test_read_not_json_serializable():
...
-    match = "datatype 'complex' of column 'a' is not in allowed values"
-    with pytest.raises(ValueError, match=match):
-        Table.read(txt, format='ascii.ecsv')
+fail
+[3,4]"""
+    with pytest.warns(InvalidEcsvDatatypeWarning,
+                      match="unexpected datatype 'object' of column 'a' is not in default datatypes"):
+        t = Table.read(txt, format='ascii.ecsv')
+    assert t['a'][0] == "fail"
+    assert type(t['a'][1]) is str
```
Comment (taldcroft) [inline astropy/io/ascii/tests/test_ecsv.py:647]: Yeah, so this is part of the original rationale for being strict, namely that this "object" type doesn't really round trip in the expected way. Anyway, the warning should do the trick.

File: astropy/io/ascii/ecsv.py [hunk 2]
```diff
@@ -175,13 +180,13 @@ def get_cols(self, lines):
...
-            # For back-compatibility with those existing older files, allow reading with no error.
-            if col.dtype not in ECSV_DATATYPES and self.ecsv_version > (0, 9, 0):
-                raise ValueError(f'datatype {col.dtype!r} of column {col.name!r} '
-                                 f'is not in allowed values {ECSV_DATATYPES}')
+            # Warn if col dtype is not a valid ECSV datatype, but allow reading for
+            # back-compatibility with existing older files that have numpy datatypes
+            # like datetime64 or object or python str, which are not in the ECSV standard.
+            if col.dtype not in ECSV_DATATYPES:
+                msg = (f'unexpected datatype {col.dtype!r} of column {col.name!r} '
+                       f'is not in default datatypes {ECSV_DATATYPES}, using anyway.')
```
Comment (taldcroft) [inline astropy/io/ascii/ecsv.py:188]: The "unexpected results" mostly applies to the `object` type but I think it's a good idea to call out that they need to be careful.
```suggestion
                       f'is not in allowed ECSV datatypes {ECSV_DATATYPES}. '
                       'Using anyway as a numpy dtype but beware since unexpected results '
                       'are possible.)
```

**Review by taldcroft — COMMENTED** [review]
@emirkmo - thanks, looks good with just minor comments.

**Comment by emirkmo** [comment]
I implemented your minor comments and now all tests are passing @taldcroft. I'm going to squash as well, then I guess it's ready to merge.

**Comment by emirkmo** [comment]
Is there anything new that needs to be done in terms of documentation? Also, I will fix the merge conflict.

**Review by taldcroft — APPROVED** [review]
Looks good @emirkmo , thanks!!

**Comment by taldcroft** [comment]
@emirkmo - ping me after you fix the merge conflict.

**Comment by emirkmo** [comment]
@taldcroft Merge conflicts resolved via rebase.

**Review by taldcroft — APPROVED** [review]
Thanks @emirkmo !
